types/mithril/index.d.ts

```diff
@@ -208,7 +208,7 @@ declare namespace Mithril {
 	 * Any Javascript object that has a view method can be used as a Mithril component.
 	 * Components can be consumed via the m() utility.
 	 */
-	interface Component<Attrs, State extends Lifecycle<Attrs, State>> extends Lifecycle<Attrs, State> {
```

Comment:
I think the real solution to your problem might be [`ThisType` introduced in TS 2.3](https://sandersn.github.io/manual/Why-ThisType.html).  But it is designed for use with type inference and contextual typing, and I didn't notice the mithril tests using type inference, so it may not work either.

Either way, I don't think constraints are useful here. Constraints do nothing but reject type arguments that are not assignable to them. A constraint doesn't add anything to the constrained type parameter once it leaves  the constraint's scope.

I tried to figure it out what the constraint was currently doing by looking at the test. Here's what I have so far:

1. *If* `oninit` (or others) is provided, it must be assignable to `(this: State, vnode: Vnode<Attrs, State>) => any` (most of these tests use it this way)
2. Instances of classes that implement `ClassComponent` should have `oninit` et al available to call (if defined) (inferred from test/test-class-component.ts)
3. (Maybe?) make sure that all properties are declared in `State` since `State extends this` isn't legal.

If the aim is (1), it's already checked without the constraint.
If the aim is (2), it's already done by `class Component extends Lifecycle`.
If the aim is to manage `this` references inside the body of `Comp` methods, then `ThisType` may be the solution. Constraints aren't doing anything here either.